### PR TITLE
commitlog_test: Add unit test for footprint + ensure disk limit >= 2 segments (per shard)

### LINKED ISSUE #9369: Commitlog disk limit should enforce >= 2 segments worth
Commitlog disk limits (and thus thresholds etc) is set from input config, as is (potentially) segment sizes. However, if a user inputs values such that we (per shard) end up with less than 2*segment size per CL shard instance, we might end up having severe problems maintaining segment juggling, at least with hard limits.

Can't really make this a check + throw, because the user would have to calculate per-shard values and multiply. In fact, even defaults could (in an insane world) be unsatisfactory (since based on avail mem vs. shard count). So should just adjust it upwards.

### PR DESCRIPTION
**Pull request by elcallio**
Refs /scylladb/scylla-dtest#2277
Fixes #9369    

A better place to verify we adhere to disk limits (as such as we do) is in a unit test, where we can get the actual per-shard limits and current actual footprint (as CL sees it).

Included is a small sanity change for commitlog init that makes disk limit max(2*segment_size, cfg.per_shard_disk_limit) to ensure 
we get workable ranges even if a user (or more likely - in this case and always - a clumsy test writer).
Not making this a check + throw for one single reason: It is potentially difficult for a user to put in a value that is sanity checked for different configs, since we need to verify this per-shard, not the actual input values.

### CODE REVIEW

**Comment by bhalevy** [comment]
Is this PR still required and valid?

File: db/commitlog/commitlog.cc [hunk 1]
```diff
@@ -1206,7 +1206,7 @@ db::commitlog::segment_manager::segment_manager(config c)
     }())
     , max_size(std::min<size_t>(std::numeric_limits<position_type>::max() / (1024 * 1024), std::max<size_t>(cfg.commitlog_segment_size_in_mb, 1)) * 1024 * 1024)
     , max_mutation_size(max_size >> 1)
-    , max_disk_size(size_t(std::ceil(cfg.commitlog_total_space_in_mb / double(smp::count))) * 1024 * 1024)
+    , max_disk_size(std::max(2 * max_size, size_t(std::ceil(cfg.commitlog_total_space_in_mb / double(smp::count))) * 1024 * 1024))
```
Comment (bhalevy) [inline db/commitlog/commitlog.cc:1209]: How is this related to the added unit test?

File: test/boost/commitlog_test.cc [hunk 2]
```diff
@@ -947,3 +947,73 @@ SEASTAR_TEST_CASE(test_commitlog_deadlock_with_flush_threshold) {
...
+    tmpdir tmp;
+    cfg.commit_log_location = tmp.path().string();
+    auto log = co_await commitlog::create_commitlog(cfg);
+
+    rp_set rps;
+    std::deque<rp_set> queue;
+    size_t n = 0;
+
+    // uncomment for verbosity
+    logging::logger_registry().set_logger_level("commitlog", logging::log_level::debug);
```
Comment (bhalevy) [inline test/boost/commitlog_test.cc:977]: This can be done externally when running the test.
Shouldn't be hard coded.

File: test/boost/commitlog_test.cc [hunk 3]
```diff
@@ -947,3 +947,73 @@ SEASTAR_TEST_CASE(test_commitlog_deadlock_with_flush_threshold) {
...
+        auto old = std::exchange(rps, rp_set{});
+        queue.emplace_back(std::move(old));
+    });
+
+    uint64_t limit = log.disk_limit(), footprint = 0;
+
+    try {
+        for (;;) {
+            auto now = timeout_clock::now();
+            rp_handle h = co_await with_timeout(now + 2s, log.add_mutation(uuid, size, db::commitlog::force_sync::no, [&](db::commitlog::output& dst) {
```
Comment (bhalevy) [inline test/boost/commitlog_test.cc:993]: why is the timeout needed?

File: test/boost/commitlog_test.cc [hunk 4]
```diff
@@ -947,3 +947,73 @@ SEASTAR_TEST_CASE(test_commitlog_deadlock_with_flush_threshold) {
...
+                dst.fill('1', size);
+            }));
+            rps.put(std::move(h));
+        }
+    } catch (timed_out_error&) {
+        // ok...
```
Comment (bhalevy) [inline test/boost/commitlog_test.cc:999]: why is a timeout ok?

Comment (bhalevy) [reply test/boost/commitlog_test.cc:999]: who waits on `log.add_mutation` that is now running astray

File: test/boost/commitlog_test.cc [hunk 5]
```diff
@@ -947,3 +947,73 @@ SEASTAR_TEST_CASE(test_commitlog_deadlock_with_flush_threshold) {
...
+    }
+
+    // we should be at max size now.
+    footprint = log.disk_footprint();
+    std::cerr << queue.size() << std::endl;
```
Comment (bhalevy) [inline test/boost/commitlog_test.cc:1004]: please use testlog for logging

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:1209]: It is not. See second paragraph of header/title.

File: test/boost/commitlog_test.cc (hunk 4 above)
Comment (elcallio) [reply test/boost/commitlog_test.cc:999]: Shutdown, below. It will not complete until any allocs are done/locked out (request_controller).

File: test/boost/commitlog_test.cc (hunk 3 above)
Comment (elcallio) [reply test/boost/commitlog_test.cc:993]: To ensure we reach a point when an allocation cannot complete.

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (bhalevy) [reply db/commitlog/commitlog.cc:1209]: I think that rather than silently trimming the value to (twice) the max, it'd be better to throw an exception with the details and refuse to start.

Comment (elcallio) [reply db/commitlog/commitlog.cc:1209]: That would break/change behaviour. I think maybe adding a warning about the threshold being upped in "from_config" (the way scylla options become cl options) is better.

File: db/commitlog/commitlog.cc [hunk 6]
```diff
@@ -96,13 +96,19 @@ class db::cf_holder {
     virtual void release_cf_count(const cf_id_type&) = 0;
 };
 
+static size_t adjusted_segment_size_in_mb(size_t s) {
+    return std::min<size_t>(std::numeric_limits<db::position_type>::max() / (1024 * 1024), std::max<size_t>(s, 1));
```
Comment (bhalevy) [inline db/commitlog/commitlog.cc:100]: why not take the minimum into account here?

e.g.
```c++
static size_t adjusted_segment_size_in_mb(size_t s, size_t min_size, size_t max_size = ) {
    return std::min<size_t>(std::numeric_limits<db::position_type>::max() / (1024 * 1024), std::max<size_t>(s, min_size));
}
```

File: db/commitlog/commitlog.cc [hunk 7]
```diff
@@ -113,6 +119,14 @@ db::commitlog::config db::commitlog::config::from_db_config(const db::config& cf
         c.commitlog_flush_threshold_in_mb = cfg.commitlog_flush_threshold_in_mb();
     }
 
+    if (c.commitlog_segment_size_in_mb < cfg.commitlog_segment_size_in_mb()) {
+        clogger.warn("Segment size {} MB too high, adjusted to {} MB", cfg.commitlog_segment_size_in_mb(), c.commitlog_segment_size_in_mb);
+    }
+    auto min_size = 2 * smp::count * c.commitlog_segment_size_in_mb;
+    if (cfg.commitlog_total_space_in_mb() >= 0 && c.commitlog_total_space_in_mb < min_size) {
```
Comment (bhalevy) [inline db/commitlog/commitlog.cc:126]: I don't understand this logic.

if `min_size = 2 * smp::count * c.commitlog_segment_size_in_mb` then `c.commitlog_segment_size_in_mb < min_size` always.

**Review by bhalevy — CHANGES_REQUESTED** [review]

File: db/commitlog/commitlog.cc (hunk 6 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:100]: We don't really have a minimum segment size. At least not defined.

File: db/commitlog/commitlog.cc (hunk 7 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:126]: Yes. But that is not the if above. It checks `commitlog_total_space_in_mb < min_size`.

File: db/commitlog/commitlog.cc (hunk 6 above)
Comment (bhalevy) [reply db/commitlog/commitlog.cc:100]: Well, 1 seems to be the minimum.
I had in mind `min_size = 2 * smp::count * c.commitlog_segment_size_in_mb` below, but I see now that it relates to the `commitlog_total_space_in_mb`, not to `commitlog_segment_size_in_mb`.

For symmetry reasons, why not have an `adjusted_total_space_in_mb` function too?

File: db/commitlog/commitlog.cc (hunk 7 above)
Comment (bhalevy) [reply db/commitlog/commitlog.cc:126]: Opps, I see.

But I don't see it adjusted in this block, just warned about.

**Comment by mykaul** [comment]
@elcallio - is this PR still relevant?

**Comment by eliransin** [comment]
@elcallio is the issue it fixes also not relevant anymore? (#9369 )

**Comment by elcallio** [comment]
@eliransin It might be. Not sure what limits/restrictions we actually want to set. Might retroactively agree with @bhalevy that instead of adjusting sizes, we should just throw iff limits are set to low.

**Comment by avikivity** [comment]
Missing backport label paperwork

**Comment by regevran** [comment]
@bhalevy - this PR is ready and needs your approval.

**Comment by mykaul** [comment]
> @bhalevy - this PR is ready and needs your approval.

There are (too?) many unresolved conversations here.

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:1514]: Warning added.

File: test/boost/commitlog_test.cc (hunk 4 above)
Comment (elcallio) [reply test/boost/commitlog_test.cc:1873]: done

File: test/boost/commitlog_test.cc (hunk 5 above)
Comment (elcallio) [reply test/boost/commitlog_test.cc]: done

File: test/boost/commitlog_test.cc (hunk 2 above)
Comment (elcallio) [reply test/boost/commitlog_test.cc]: removed

**Comment by elcallio** [comment]
rebased
